Fourier basis

zbmorrow opened the pull request:
This branch implements sparse Fourier interpolation. The syntax of `GridFourier::makeGrid` is

```c++
void GridFourier::makeGrid(int num_dimensions, int num_outputs, int depth, const int *anisotropic_weights = 0, const int *level_limits = 0)
```
similarly to the syntax of `GridGlobal::makeGrid`. The other user-facing commands (`loadNeededPoints`, `evaluate`, `evaluateBatch`, etc.) behave the same as in the `GridGlobal` class.

The one-dimensional Fourier grid on the canonical domain [0,1] has 3^_l_ (_l_>=0) equally spaced points
<img src="https://latex.codecogs.com/gif.latex?x_k&space;=&space;\frac{k}{3^\ell},&space;\hspace{1em}&space;k=0,1,\dots,3^\ell-1," title="x_k = \frac{k}{3^\ell}, \hspace{1em} k=0,1,\dots,3^\ell-1," />
and the set of one-dimensional basis functions on level _l_ is

<img src="https://latex.codecogs.com/gif.latex?\left\{&space;\omega_k(x)&space;=&space;e^{2\pi&space;i&space;k&space;x}&space;~:~k=0,\pm1,\pm2,&space;\dots,&space;\pm\frac{3^\ell-1}{2}\right\}" title="\left\{ \omega_k(x) = e^{2\pi i k x} ~:~k=0,\pm1,\pm2, \dots, \pm\frac{3^\ell-1}{2}\right\}" />

where <img src="https://latex.codecogs.com/gif.latex?i=\sqrt{-1}" title="i=\sqrt{-1}" />. Note that the sets of nodes and basis functions are nested with respect to _l_.

In order to construct the interpolation coefficients, we must perform a discrete Fourier transform of the y-data (in each output direction) on the constituent tensor-product grids. This is done once, when the function values are loaded at the front end, and is decoupled from the evaluation of the complex exponential basis functions, which must be done for each evaluation point. The DFT is implemented in `TasmanianFourierTransform::dft` within `tsgLinearSolvers.cpp`.

dalg24-jenkins commented:
Can one of the admins verify this patch?

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,91 @@
+
+namespace TasGrid{
+
+class GridFourier{
```
* must inherit from `BaseCanonicalGrid`
* must also add multiple `evaluate()` functions, you can make those fallback to basic `evaluate()` and `evaluateBatch()`

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,91 @@
+    double* getPoints() const;
+    void getPoints(double *x) const; // returns the loaded points unless no points are loaded, then returns the needed points
+
+	void calculateFourierCoefficients();
```
* tab for indentation, use spaces

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,91 @@
+
+	void calculateFourierCoefficients();
+
+    std::complex<double>* getBasisFunctions(const double x[]) const;
```
* shouldn't those be `protected`?

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,91 @@
+    void evaluate(const double x[], double y[]) const;
+    void evaluateBatch(const double x[], int num_x, double y[]) const;
+
+	void clearAccelerationData();
```
* check indentation across all files, there are many places where tabs are used instead of spaces

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,91 @@
+	std::complex<double> *fourier_coefs;
+    int **exponent_refs;
+
+    int *max_levels; // for evaluation purposes, counts the maximum level in each direction (only counts tensors)
```
* is this used, `GridFourier::evaluate()` doesn't seem to need this.

mkstoyanov commented on SparseGrids/tsgLinearSolvers.cpp:
```diff
@@ -167,6 +167,76 @@ void TasmanianTridiagonalSolver::decompose(int n, double d[], double e[], double
+    std::complex<double>* exp_k = new std::complex<double>[rank];
+    std::fill(exp_k, exp_k+rank, 1);
+
+    for(int k=0; k<num_data; k++) {
```
* this is the $O(n^2)$ algorithm, right? If so, we need to think about this.

zbmorrow replied:
Yes, the algorithm I wrote is $O(n^2)$ (one-time at the front end, but could still be bad for grids with large depth).

mkstoyanov commented on SparseGrids/tsgOneDimensionalWrapper.cpp:
```diff
@@ -183,6 +183,8 @@ OneDimensionalWrapper::OneDimensionalWrapper(const OneDimensionalMeta *meta, int
             GreedySequences greedy;
             unique = greedy.getMinDeltaNodes(meta->getNumPoints(max_level, rule));
+        }else if (rule==rule_fourier) {
+            unique = OneDimensionalNodes::getFourierNodes(max_level);
```
* wrapper is missing quadrature weights for integration

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* memory leak here, this needs to be `delete[]`-ed

mkstoyanov commented on SparseGrids/tasgrid_main.cpp:
```diff
@@ -156,6 +156,7 @@ int main(int argc, const char ** argv){
             if ((strcmp(argv[k],"global") == 0)) test = test_global;
             if ((strcmp(argv[k],"local") == 0)) test = test_local;
             if ((strcmp(argv[k],"wavelet") == 0)) test = test_wavelet;
+            if ((strcmp(argv[k],"fourier") == 0)) test = test_fourier;
```
* looks much better
* you need to add a test to `SparseGrids/CMakeLists.txt` with cmake command `add_test()`
* then we can test on the CI

```
add_test(SparseGridsFourier      tasgrid -test fourier)
```

zbmorrow replied:
Done

mkstoyanov commented:
OK to test

zbmorrow commented:
Updates:

- Tests for Fourier rule, acceleration, and domain transform (refinement omitted for now)
- Tests incorporated into `tasgridExternalTests` (added a couple new functions to `tasgridTestFunctions`); tests also run upon install
- Added BLAS for complex matvec/matmat
- Implemented `getInterpolationWeights` for nodal interpolation (does DFT of complex exponentials on each tensor)
- Implemented quadrature rule
- Inherits from `BaseCanonicalGrid` class and is fully implemented in `TasmanianSparseGrid` class

Still needed:
- IO for binary/ASCII
- O(N log N) Fourier transform algorithm
- Is there a nicer way to re-index when reordering the data for the Fourier transform?

mkstoyanov commented:
@zbmorrow: check the comments, there is something wrong with the BLAS fall-back functions

mkstoyanov commented:
@zbmorrow, so far so good.  

* Next, we need to rebase to current master to fix the conflicts, then I'll go over the code again to see if I catch anything or if we have to massage the internals.
* We will work on more tests and performance afterwards.

mkstoyanov commented on SparseGrids/tasgridTestFunctions.cpp:
```diff
@@ -80,6 +80,14 @@ TwoOneCosCos::TwoOneCosCos(){} TwoOneCosCos::~TwoOneCosCos(){} int TwoOneCosCos:
+
+TwoTwoSinCos::TwoTwoSinCos() {} TwoTwoSinCos::~TwoTwoSinCos() {} int TwoTwoSinCos::getNumInputs() const{ return 2; } int TwoTwoSinCos::getNumOutputs() const{ return 2; }
+const char* TwoTwoSinCos::getDescription() const{ return "f(x,y) = [sin(2*pi*x)cos(4*pi*y), cos(2*pi*x)]"; }
+void TwoTwoSinCos::eval(const double x[], double y[]) const{ y[0] = sin(2*M_PI*x[0])*cos(4*M_PI*x[1]); y[1]=cos(2*M_PI*x[0]); } void TwoTwoSinCos::getIntegral(double y[]) const{ y[0] = 0.0; y[1]=0.0; }
```
* `cos(2 * M_PI * x)` is too simple, try something that doesn't belong to the basis

mkstoyanov commented on SparseGrids/tasgridExternalTests.cpp:
```diff
@@ -89,13 +91,14 @@ TestResults ExternalTester::getError(const BaseFunction *f, TasGrid::TasmanianSp
     if ((type == type_integration) || (type == type_nodal_interpolation)){
-        double *points = grid->getPoints(), *weights = 0;
+        double *points = grid->getPoints();
+        double *weights = new double[num_points];
```
Why assigning weights here? you are overwriting the pointer below.

zbmorrow replied:
Reverted. I did this while trying to isolate some weird behavior, and I forgot to change it back.

mkstoyanov commented on SparseGrids/tsgCoreOneDimensional.cpp:
```diff
@@ -217,7 +217,7 @@ OneDimensionalMeta::~OneDimensionalMeta(){}
 
 int OneDimensionalMeta::getNumPoints(int level, TypeOneDRule rule) const{
-    int lcc;
+    int lcc, fourier_result;
```
You can reuse `lcc`, no need for another variable here.

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* Did you fix this, or am I reading something old?

zbmorrow replied:
Huh, could have sworn I fixed this... I screwed up halfway through the rebase yesterday, which could be the problem. Fixed now at any rate.

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* no need for a variable
```
double *x = new double[num_dimensions * points->getNumIndexes()];
```

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* we will have to fix the indexing and the fast algorithm later

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* isn't this the same as: `evaluateHierarchicalFunctions()`

zbmorrow replied:
`evaluateHierarchicalFunctions()` gives the weights as an array of `double`s (real and imag parts), rather than complex numbers. Should I park this code in an overloaded function that gives the complex weights?

mkstoyanov replied:
Here you have the exact same loop as in `evaluateHierarchicalFunctions()`, just call the function, why write the loop twice?

zbmorrow replied:
Done

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* Why skip OpenMP here?

mkstoyanov commented on SparseGrids/tsgGridFourier.cpp:
* `TasmanianSparseGrids` class has `enabled/disabled` functions, what is the purpose of this function here?

mkstoyanov commented on SparseGrids/tsgGridGlobal.cpp:
```diff
@@ -916,13 +916,13 @@ double* GridGlobal::computeSurpluses(int output, bool normalize) const{
         double *w = gg->getQuadratureWeights();
         double *x = gg->getPoints();
-        double *I = new double[qn];
+        double *I_quad = new double[qn];
```
* Is this change still needed?

zbmorrow replied:
Nope

mkstoyanov commented on SparseGrids/tsgGridFourier.hpp:
```diff
@@ -0,0 +1,117 @@
+
+#include <cstdlib>
+#include <math.h>
+#include <complex>
```
* Is it necessary to include the `<complex>` header here? 
* Can this be hidden in the `.cpp` implementation without propagating to the user?

zbmorrow replied:
When I move it to the `.cpp` and attempt to `make`, I get errors because the functions in the `.hpp` files have inputs of type `std::complex<typename>`.

mkstoyanov replied:
* Yes, the question pertains to whether it is possible to redesign the header so that you avoid including `<complex>`

mkstoyanov commented on SparseGrids/tsgHiddenExternals.hpp:
```diff
@@ -30,6 +30,8 @@
 #ifndef __TASMANIAN_SPARSE_GRID_HIDDEN_INTERNALS_HPP
 #define __TASMANIAN_SPARSE_GRID_HIDDEN_INTERNALS_HPP
+#include <complex>
+using std::complex;
```
* *Never* force a namespace like that, all headers that include this would now work in the `std::complex` namespace
* if you need a variable from `std::complex`, then use that explicitly
```
using std::complex <some symbol>
```
* sometimes is it better to keep explicitly including from the namespace rather than contaminating your list of symbols

mkstoyanov commented on SparseGrids/tsgHiddenExternals.hpp:
```diff
@@ -41,8 +43,10 @@ extern "C" double ddot_(const int *N, const double *x, const int *incx, const do
 extern "C" double dnrm2_(const int *N, const double *x, const int *incx);
 extern "C" void dgemv_(const char *transa, const int *M, const int *N, const double *alpha, const double *A, const int *lda, const double *x, const int *incx, const double *beta, const double *y, const int *incy);
 extern "C" void dtrsv_(const char *uplo, const char *trans, const char *diag, const int *N, const double *A, const int *lda, const double *x, const int *incx);
+extern "C" void zgemv_(const char *transa, const int *M, const int *N, const complex<double> *alpha, const complex<double> *A, const int *lda, const complex<double> *x, const int *incx, const complex<double> *beta, const complex<double> *y, const int *incy);
```
* isn't this the same as:
```
extern "C" ... const int *N, const double *alpha, ...
```
with the added note that alpha is 2 doubles

mkstoyanov commented on SparseGrids/tsgLinearSolvers.cpp:
```diff
@@ -167,6 +167,74 @@ void TasmanianTridiagonalSolver::decompose(int n, double d[], double e[], double
+        num_data *= n[r];
+    }
+
+    std::complex<double> i(0,1);
```
`i(0.0, 1.0)`

* With templates, you must be extra explicit when declaring thing, the compiler has to guess lots of stuff already and the chance for error is increased.

mkstoyanov commented on SparseGrids/TasmanianSparseGrid.cpp:
```diff
@@ -220,6 +221,17 @@ void TasmanianSparseGrid::makeWaveletGrid(int dimensions, int outputs, int depth
+    fourier = new GridFourier();
+    fourier->makeGrid(dimensions, outputs, depth, type, anisotropic_weights, level_limits);
+    base = fourier;
+    if(level_limits != 0) {
```
* The code must be formatter to be consistent with the rest (as well as self-consistent)
    * sometimes you leave a space between `) {`, do not
    * one space after `if (`
    * spaces around `=`, e.g., `i = 1`, there are exceptions here but only when trying to save space and make the code compact
    * there are places with multiple empty lines between functions, use one or zero lines
    * remove all trailing spaces and empty lines

zbmorrow replied:
I believe I have taken care of items 1, 2, 4, and 5.  

For item 3, I did a reg exp search of (at least most of) the documents where I've made changes. The only instances I saw were in `for` loops. Was there a place in particular where I needed to fix the spaces around `=`?

The regexp I used was `[^( |=|(for\(int (a-z)))]=[^( |=)]`

mkstoyanov replied:
The spaces around `=` go usually outside of the loop, I have to go back to find the spot where I saw it. Look over for code such as
```
base=fourier;
```

zbmorrow replied:
I think I found it in `tasgridTestFunctions.cpp`.

zbmorrow commented on SparseGrids/tasgridExternalTests.cpp:
```diff
@@ -933,8 +936,8 @@ bool ExternalTester::testAllWavelet() const{
     }else{
-        cout << setw(wfirst) << "Rule" << setw(wsecond) << TasGrid::OneDimensionalMeta::getIORuleString(rule_wavelet) << setw(wthird) << "FAIL" << endl; pass = false;
-    }{ TasGrid::TasmanianSparseGrid grid;
+        cout << setw(wfirst) << "Rules" << setw(wsecond) << TasGrid::OneDimensionalMeta::getIORuleString(rule_wavelet) << setw(wthird) << "FAIL" << endl; pass = false;
```
It was a small thing I noticed. The line in the success part of the `if` statement has `Rules`. Based on the tests for global, sequence, etc., it looked as though when the tests for all rules of a given grid class pass, the program prints `Rules so-and-so Pass` or `Rules so-and-so FAIL`. I'll undo it in this branch since it's completely unrelated to the Fourier stuff.

mkstoyanov replied:
* Global and Sequence relate to multiple `Rules`, hence each rule has a `Rule` pass/fail and the group has a group `Rules` pass/fail
* Wavelet rules have two variants, even though it is technically the same name
* Fourier has not variants (yet), but keep it the same as the rest

mkstoyanov commented on SparseGrids/tasgridWrapper.cpp:
```diff
@@ -1048,6 +1097,8 @@ bool TasgridWrapper::executeCommand(){
     }else if (command == command_makewavelet){
         createWaveletGird();
+    }else if (command == command_makefourier){
+        createFourierGird();
```
* see earlier comment about merging this with `makeSequenceGrid()`

mkstoyanov commented on SparseGrids/tsgLinearSolvers.cpp:
```diff
@@ -167,6 +167,68 @@ void TasmanianTridiagonalSolver::decompose(int n, double d[], double e[], double
+        num_data *= n[r];
+    }
+
+    std::complex<double> i(0.0, 1.0);
```
If the complex constant is defined in `std::complex::I`, then there is no need to redefine it here, just use it. I think it will work if you just say:
```
using std::complex::I;
```
and then use `I` afterwards.

zbmorrow replied:
I get an error with `std::complex::I`. I looked on Stackexchange but couldn't find `sqrt(-1)` hardcoded somewhere as a constant within `std::complex` in C++11.

mkstoyanov replied:
@zbmorrow I thought we had a problem with `I` being declared as `sqrt(-1)` and that's why you had to remove the `I` that was used for integral. Then the problem went away when we stopped using `using namespace std::complex`

zbmorrow replied:
That was from `<complex.h>`, which defined `I` as a macro (http://pubs.opengroup.org/onlinepubs/009695399/basedefs/complex.h.html). The problem stopped when we changed `<complex.h>` to `<complex>`.

zbmorrow commented on SparseGrids/tasgridWrapper.cpp:
```diff
@@ -424,10 +463,10 @@ void TasgridWrapper::outputHierarchicalCoefficients() const{
     int num_d = grid->getNumOutputs();
     if (outfilename != 0){
-        writeMatrix(outfilename, num_p, num_d, coeff, useASCII);
+        writeMatrix(outfilename, num_p, num_d, coeff, useASCII, grid->isFourier());
```
Why not handle `printMatrix` the same as `writeMatrix`?
```
printMatrix(num_p, (grid->isFourier() ? 2 * num_d : num_d), coeff);
```

mkstoyanov replied:
Because `printMatrix()` is human readable and the format should be
```
<number> + <number> i  <number> + <number> i ...
```
I must have missed to add the comment regarding `printMatrix()`

mkstoyanov commented:
* close/reopen to retest

mkstoyanov commented on SparseGrids/tasgridExternalTests.cpp:
```diff
@@ -1333,6 +1411,21 @@ bool ExternalTester::testAllDomain() const{
         }
+    }{
+        const BaseFunction *f = &f21coscos;
+        double errs[5] = {5.E-2, 7.E-3, 9.E-4, 1.E-13};
```
* formatting, add spaces after `{` and before `}`

mkstoyanov commented on SparseGrids/tasgridWrapper.cpp:
```diff
@@ -1026,13 +1059,21 @@ void TasgridWrapper::writeMatrix(const char *filename, int rows, int cols, const
+        if (isComplex){
+            for(size_t j=0; j<((size_t) cols); j++){
+                cout << setw(25) << mat[i* ((size_t) cols) + ((size_t) 2*j)];   // real part
+                cout << (mat[i* ((size_t) cols) + ((size_t) 2*j+1)] < 0 ? " -" : " +");
```
* use `< 0.0`, i.e., specify that you are comparing floats

mkstoyanov commented on SparseGrids/tasgridWrapper.cpp:
```diff
@@ -1026,13 +1059,21 @@ void TasgridWrapper::writeMatrix(const char *filename, int rows, int cols, const
+            for(size_t j=0; j<((size_t) cols); j++){
+                cout << setw(25) << mat[i* ((size_t) cols) + ((size_t) 2*j)];   // real part
+                cout << (mat[i* ((size_t) cols) + ((size_t) 2*j+1)] < 0 ? " -" : " +");
+                cout << setw(24) << abs(mat[i* ((size_t) cols) + ((size_t) 2*j+1)]) << " i   ";     // imag part
```
* this is a bug, `abs()` works only for integer types, long, int, etc.
* you must use `fabs()` for floating point numbers
* at the end, why 2 spaces as opposed to 1?
* why not have `i` attached to the number, e.g., `1.234E-5 + 1.234.E-5i`

mkstoyanov commented on SparseGrids/tsgCoreOneDimensional.cpp:
```diff
@@ -1056,6 +1067,23 @@ double* OneDimensionalNodes::getRLejaShifted(int n){
+    if(level > 0) { x[1]=1.0/3.0; x[2]=2.0/3.0; }
+    int count=3;
+    for(int l=2; l<=level; l++) {
+        n = meta.getNumPoints(l,rule_fourier);
```
* formatting, add space after the comma: `(l, rule_fourier)`

mkstoyanov commented on SparseGrids/tsgLinearSolvers.cpp:
```diff
@@ -167,6 +167,69 @@ void TasmanianTridiagonalSolver::decompose(int n, double d[], double e[], double
     }
 }
 
+void TasmanianFourierTransform::dft(const int rank, const int n[], const std::complex<double> data[], std::complex<double> out[], const int direction){
```
* rename the `dft` function to something like: `discrete_fourier_transform`

if you google `dft code` the top hits are related to Density Field Theory, which is a huge field. Nobody is really using the simple Fourier transfrom, and we will drop it soon.

mkstoyanov commented on SparseGrids/tsgLinearSolvers.cpp:
```diff
@@ -167,6 +167,69 @@ void TasmanianTridiagonalSolver::decompose(int n, double d[], double e[], double
+    // Fourier transform code by ZBM
+    // DOES NOT NORMALIZE the outgoing Fourier coefficients
+
+    if (direction != 1 && direction != -1) std::cerr << "ERROR: direction must be -1 (forward) or 1 (backward); value is " << direction << std::endl;
```
* do not use simple integers as flags in C++, it is OK for C or Fortran, but C++ has more elegant ways, e.g., boolean or enumerate both of which will make the code more readable
* do not change this now, the whole function will be scrapped anyway

mkstoyanov commented:
* added Fourier grids that are based on complex exponential basis
